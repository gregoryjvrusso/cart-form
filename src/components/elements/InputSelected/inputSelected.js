import React from "react";
import Box from "../Box";
import { arrayOf, func, shape, string, object } from "prop-types";
import { Container, Option, Label, TextMessage } from "./inputSelected.styles";

const InputSelected = ({
  name,
  textLabel,
  onInputChange,
  options,
  messageError,
  value,
  setValue,
  error,
  setError,
  onInputBlur,
  ...props
}) => (
  <Box width={"100%"} pt={3}>
    <Container
      name={name}
      placeholder={textLabel}
      value={value}
      required
      data-testid="input-select"
      onBlur={(e) => onInputBlur(e, error, setError, name)}
      onChange={(e) => onInputChange(e, setValue, value, name)}
      {...props}
    >
      {options.map((e) => (
        <Option data-testid="options" key={e.value} value={e.value}>
          {e.text}
        </Option>
      ))}
    </Container>
    <Label forHtml={name} {...props}>
      {textLabel}
    </Label>
    {error[name] && <TextMessage>{messageError}</TextMessage>}
  </Box>
);

InputSelected.propTypes = {
  name: string.isRequired,
  textLabel: string.isRequired,
  messageError: string,
  onInputChange: func.isRequired,
  options: arrayOf(
    shape({
      value: string,
      text: string,
    })
  ),
  value: string,
  setValue: func,
  error: object,
  setError: func,
  onInputBlur: func,
};

InputSelected.defaultProps = {
  messageError: "",
  options: [],
  value: "",
  error: {},
  setValue: undefined,
  setError: undefined,
  onInputBlur: undefined,
};

InputSelected.displayName = "InputSelected";

export default InputSelected;
