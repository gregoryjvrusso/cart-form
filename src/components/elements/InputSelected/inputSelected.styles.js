import { theme } from "styled-tools";
import styled, { css } from "styled-components";
import { arrowBottom } from "../../../temp";
import Text from "../Text";

export const Container = styled.select`
  border: 0;
  border-bottom: 2px solid
    ${(props) =>
      props.error ? theme(`colors.texts.red`) : theme(`colors.texts.gray`)};
  background-color: transparent;
  width: 100%;
  font-size: 17px;
  user-select: none;
  appearance: none;
  background: url(${arrowBottom}) no-repeat center right;
  &::placeholder {
    color: transparent;
  }
  ${(props) =>
    props.value &&
    css`
      ~ label {
        top: 6px;
        left: 5px;
        font-size: 13px;
        opacity: 0.6;
      }
    `}
`;

export const Option = styled.option`
  height: 20px;
  font-size: 15px;
  background: transparent;
  color: ${theme(`colors.texts.gray`)};
`;

export const Label = styled.label`
  position: absolute;
  pointer-events: none;
  left: 10px;
  top: 17px;
  transition: 0.3s ease all;
  font-size: 17px;
  color: ${(props) =>
    props.error ? theme(`colors.texts.red`) : theme(`colors.texts.gray`)};

  ${Container}:focus ~ & {
    top: 6px;
    left: 5px;
    font-size: 13px;
    opacity: 0.6;
  }
`;

export const TextMessage = styled(Text)`
  font-size: 13px;
  color: ${theme(`colors.texts.red`)};
  margin-top: 0;
`;
