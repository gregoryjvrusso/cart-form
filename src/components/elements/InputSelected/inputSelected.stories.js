import React, { useState } from "react";
import { storiesOf } from "@storybook/react";
import { ThemeProvider } from "styled-components";
import theme from "../../../common/theme";

import InputSelected from "../InputSelected";

const mockData = [
  {
    value: "",
    text: "",
  },
  {
    value: "1",
    text: "Parcelar em 1 vez",
  },
  {
    value: "2",
    text: "Parcelar em 2 vezes",
  },
  {
    value: "3",
    text: "Parcelar em 3 vezes",
  },
  {
    value: "4",
    text: "Parcelar em 4 vezes",
  },
  {
    value: "5",
    text: "Parcelar em 5 vezes",
  },
  {
    value: "6",
    text: "Parcelar em 6 vezes",
  },
];

storiesOf("Elements/InputSelected", module).add("default", () => {
  const [value, setValue] = useState("");
  const [error, setError] = useState(false);
  const mockOnInputChange = (e) => {
    setValue(e.target.value);
    e.target.value ? setError(false) : setError(true);
  };
  return (
    <ThemeProvider theme={theme}>
      <InputSelected
        onInputChange={mockOnInputChange}
        value={value}
        name="qntParcelas"
        options={mockData}
        textLabel="Quantidade de parcelas"
        error={error}
        messageError="Insira o nÃºmero de parcelas"
      />
    </ThemeProvider>
  );
});
