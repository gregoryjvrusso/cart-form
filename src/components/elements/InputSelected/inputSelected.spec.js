import React from "react";
import { render } from "@testing-library/react";
import InputSelected from "../InputSelected";

const mockData = [
  {
    value: "1",
    text: "Parcelar em 1 vez",
  },
  {
    value: "2",
    text: "Parcelar em 2 vezes",
  },
  {
    value: "3",
    text: "Parcelar em 3 vezes",
  },
  {
    value: "4",
    text: "Parcelar em 4 vezes",
  },
  {
    value: "5",
    text: "Parcelar em 5 vezes",
  },
  {
    value: "6",
    text: "Parcelar em 6 vezes",
  },
];

describe("InputSelected component", () => {
  it("deve conseguir renderizar [snapshot]", async () => {
    const onChangeMock = jest.fn();
    const onBlurMock = jest.fn();
    const { container } = render(
      <InputSelected
        name="parcelas"
        onInputChange={onChangeMock}
        onInputBlur={onBlurMock}
        options={mockData}
        textLabel="Quantidade de parcelas"
        messageError="Insira o nÃºmero de parcelas"
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });
});
