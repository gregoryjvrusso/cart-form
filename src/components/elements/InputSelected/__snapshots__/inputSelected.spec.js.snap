// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputSelected component deve conseguir renderizar [snapshot] 1`] = `
<div
  class="sc-AxjAm jpulAN"
  data-testid="box"
  width="100%"
>
  <select
    class="sc-AxiKw kEbYpS"
    data-testid="input-select"
    name="parcelas"
    placeholder="Quantidade de parcelas"
    required=""
  >
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="1"
    >
      Parcelar em 1 vez
    </option>
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="2"
    >
      Parcelar em 2 vezes
    </option>
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="3"
    >
      Parcelar em 3 vezes
    </option>
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="4"
    >
      Parcelar em 4 vezes
    </option>
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="5"
    >
      Parcelar em 5 vezes
    </option>
    <option
      class="sc-AxhCb cLsAfa"
      data-testid="options"
      value="6"
    >
      Parcelar em 6 vezes
    </option>
  </select>
  <label
    class="sc-AxhUy TMHKD"
  >
    Quantidade de parcelas
  </label>
</div>
`;
